backend/src/controllers/companyController.ts

```diff
@@ -1,39 +1,166 @@
 import Company from "src/models/Company";
+import { matchedData, validationResult } from "express-validator";
 import asyncHandler from "express-async-handler";
 import createHttpError from "http-errors";
 
 // @desc Get companies matching the query
 // @route GET /api/companies
 // @access Private
 export const getCompanies = asyncHandler(async (req, res, next) => {
-  const { page, perPage, query, state } = req.body;
+  const result = validationResult(req);
+  if (!result.isEmpty()) {
+    return next(
+      createHttpError(400, result.array({ onlyFirstError: true })[0].msg),
+    );
+  }
+
+  const { page, perPage, query, state } = matchedData(req);
+
+  // Begin query
+  const dbQuery = Company.find();
+
+  // Duplicate needed to count the number of documents
+  const countQuery = Company.find();
+
+  // Search by name if provided
+  if (query !== "") {
+    countQuery.where("name").regex(new RegExp(query, "i"));
+    dbQuery.where("name").regex(new RegExp(query, "i"));
+  }
+
+  // Filter by state if provided
+  if (state !== "all") {
+    countQuery.where("state").equals(state);
+    dbQuery.where("state").equals(state);
+  }
+
+  const total = await countQuery.countDocuments().exec();
+
+  // Paginate
+  dbQuery.skip(page * perPage).limit(perPage);
+
+  const companies = await dbQuery.lean().exec();
```

Comment:
ChatGPT suggested using `Promise.all()` to execute the queries in parallel which makes it a little faster:
```
const [total, companies] = await Promise.all([
      countQuery.countDocuments().exec(),
      dbQuery.skip(page * perPage).limit(perPage).lean().exec(),
]);
```
And also maybe you could just do `countQuery = dbQuery.clone()` to remove a bit of redundant code. But tbh I dont think any of this makes too big of a difference lol

Reply:
Alright I added this lol, I ran it a few times and I think it is ever so slightly faster consistently so 🤷